File: exercise/php/day23/src/Core/Control/ControlSystem.php
<?php

namespace Core\Control;

use External\Deer\Reindeer;
use External\Stable\MagicStable;

class ControlSystem
{
    private int $XmasSpirit = 40;
    private Dashboard $dashboard;
    private MagicStable $magicStable;
    private array $reindeerPowerUnits;
    public SleighEngineStatus $status;
    public SleighAction $action;
    private float $controlMagicPower = 0;

    public function __construct()
    {
        $this->dashboard = new Dashboard();
        $this->magicStable = new MagicStable();
        $this->reindeerPowerUnits = $this->bringAllReindeers();
    }

    private function bringAllReindeers(): array
    {
        return array_map([$this, 'attachPowerUnit'], $this->magicStable->getAllReindeers());
    }

    public function attachPowerUnit(Reindeer $reindeer): ReindeerPowerUnit
    {
        return new ReindeerPowerUnit($reindeer);
    }

    public function startSystem(): void
    {
        $this->dashboard->displayStatus("Starting the sleigh...");
        $this->status = SleighEngineStatus::ON;
        $this->dashboard->displayStatus("System ready.");
    }

    /**
     * @throws ReindeersNeedRestException
     * @throws SleighNotStartedException
     */
    public function ascend(): void
    {
        if ($this->status === SleighEngineStatus::ON) {
            foreach ($this->reindeerPowerUnits as $reindeerPowerUnit) {
                $this->controlMagicPower += $reindeerPowerUnit->harnessMagicPower();
            }

            if ($this->checkReindeerStatus()) {
                $this->dashboard->displayStatus("Ascending...");
                $this->action = SleighAction::FLYING;
                $this->controlMagicPower = 0;
            } else {
                throw new ReindeersNeedRestException();
            }
        } else {
            throw new SleighNotStartedException();
        }
    }

    public function descend(): void
    {
        if ($this->status === SleighEngineStatus::ON) {
            $this->dashboard->displayStatus("Descending...");
            $this->action = SleighAction::HOVERING;
        } else {
            throw new SleighNotStartedException();
        }
    }

    public function park(): void
    {
        if ($this->status === SleighEngineStatus::ON) {
            $this->dashboard->displayStatus("Parking...");

            foreach ($this->reindeerPowerUnits as $reindeerPowerUnit) {
                $reindeerPowerUnit->getReindeer()->resetHarnessingTimes();
            }

            $this->action = SleighAction::PARKED;
        } else {
            throw new SleighNotStartedException();
        }
    }

    public function stopSystem(): void
    {
        $this->dashboard->displayStatus("Stopping the sleigh...");
        $this->status = SleighEngineStatus::OFF;
        $this->dashboard->displayStatus("System shutdown.");
    }

    private function checkReindeerStatus(): bool
    {
        return $this->controlMagicPower >= $this->XmasSpirit;
    }
}

File: exercise/scala/day23/src/main/scala/core/control/ControlSystem.scala
package core.control

import external.deer.Reindeer
import external.stable.MagicStable

class ControlSystem {
  // The Xmas spirit is 40 magic power units
  private val xmasSpirit = 40
  private val dashboard = new Dashboard()
  private val magicStable = new MagicStable()
  private val reindeerPowerUnits = bringAllReindeers()
  var status: SleighEngineStatus.Value = SleighEngineStatus.OFF
  var action: SleighAction.Value = SleighAction.PARKED
  private var controlMagicPower = 0f

  def startSystem(): Unit = {
    dashboard.displayStatus("Starting the sleigh...")
    status = SleighEngineStatus.ON
    dashboard.displayStatus("System ready.")
  }

  @throws[ReindeersNeedRestException]
  @throws[SleighNotStartedException]
  def ascend(): Unit = {
    if (status == SleighEngineStatus.ON) {
      reindeerPowerUnits.foreach(reindeerPowerUnit => {
        controlMagicPower += reindeerPowerUnit.harnessMagicPower()
      })

      if (checkReindeerStatus()) {
        dashboard.displayStatus("Ascending...")
        action = SleighAction.FLYING
        controlMagicPower = 0
      } else throw new ReindeersNeedRestException()
    } else {
      throw new SleighNotStartedException()
    }
  }

  private def checkReindeerStatus(): Boolean = controlMagicPower >= xmasSpirit

  @throws[SleighNotStartedException]
  def descend(): Unit = {
    if (status == SleighEngineStatus.ON) {
      dashboard.displayStatus("Descending...")
      action = SleighAction.HOVERING
    } else throw new SleighNotStartedException()
  }

  @throws[SleighNotStartedException]
  def park(): Unit = {
    if (status == SleighEngineStatus.ON) {
      dashboard.displayStatus("Parking...")

      reindeerPowerUnits.foreach(reindeerPowerUnit => {
        reindeerPowerUnit.reindeer.timesHarnessing = 0
      })

      action = SleighAction.PARKED
    } else throw new SleighNotStartedException()
  }

  def stopSystem(): Unit = {
    dashboard.displayStatus("Stopping the sleigh...")
    status = SleighEngineStatus.OFF
    dashboard.displayStatus("System shutdown.")
  }

  private def bringAllReindeers(): List[ReindeerPowerUnit] = {
    magicStable.getAllReindeers.map(attachPowerUnit)
  }

  def attachPowerUnit(reindeer: Reindeer): ReindeerPowerUnit = new ReindeerPowerUnit(reindeer)
}